cowbird/permissions_synchronizer.py

```diff
@@ -238,21 +269,39 @@ def _find_matching_res(self, service_type: str, resource_nametype_path: str) ->
             # An error would be raised because 2 matches of the same length would be found.
             # - /**/file
             # - /file
+            #
+            # In the case where a regex is used, the behavior is changed to search for the exact match in the res_segments.
+            # The lengh of the match is used to favor a more specific match.
+            # Example:
+            # 1:
+            # - //dir1/dir2//
+            # - //dir1/dir2//dir3// # We favor this path if it matches since it is more specific.
+            # note: It is possible to have multiple dir in the same segment when using a custom regex that extract a display_name
+            # containing a path to a target resource.
 
             matched_length_by_res = {}
             matched_groups_by_res = {}
             service_resources = self.services[service_type]
             for res_key, res_segments in service_resources.items():
                 res_regex, named_segments_count = SyncPoint._generate_regex_from_segments(res_segments)
-                matches = re.match(res_regex, resource_nametype_path)
+                resource_nametype_path = SyncPoint._generate_nametype_path_from_segments(res_segments, src_resource_tree)
+                if named_segments_count == -1:
+                    # To be able to match a path anywhere in the resource_nametype_path we need to use search
+                    # only when the field regex is passed in the res_segments. This allow to stay backward compatible.
+                    matches = re.search(res_regex, resource_nametype_path)
+                else:
+                    matches = re.match(res_regex, resource_nametype_path)
                 if matches:
-                    matched_groups = matches.groupdict()
+                    exact_match = matches.group()
+                    matched_groups = matches.groupdict() if named_segments_count != -1 else exact_match
                     if "multi_token" in matched_groups:
                         matched_groups["multi_token"] = SyncPoint._remove_type_from_nametype_path(
                             matched_groups["multi_token"]
                         )
                     matched_groups_by_res[res_key] = matched_groups
-                    matched_length_by_res[res_key] = named_segments_count
+                    # Since we want to be able to match multiple dir /dir1/dir2/dir3/** in the same segment if a custom regex is passed.
+                    # We need to use the len of the exact match to avoid matching the wrong res_key
+                    matched_length_by_res[res_key] = named_segments_count if named_segments_count != -1 else len(exact_match)
```

Comment:
Yes It would search for more segments in different res_key. Example a res key for a match of a stac collection : the segment `/stac/collections/{thredds:birdhouse/testdata/xclim/cmip6}` with the regex `[\w]+:[\w\/]+` will match `birdhouse/testdata/xclim/cmip6`, but if the permission update were made to a stac item the segment `/stac/collections/{thredds:birdhouse/testdata/xclim/cmip6/file.nc}` with the regex '[\w]+:[\w\/.-]+' would match a longer string `birdhouse/testdata/xclim/cmip6/file.nc`. That way we know we need to change the permission on a file and not in the directory.